File: examples/python/case07_test.py
import threading
import unittest

from case07 import (
    CancelSignal,
    ResultHandoff,
    blocking_send_with_cancel,
    downstream_without_cancellation,
    long_work_check_cancel,
    precheck_then_block,
)


class Case07Test(unittest.TestCase):
    def test_cancellation_at_handoff(self):
        weak_cancel, weak_handoff = CancelSignal(), ResultHandoff()
        weak = threading.Thread(target=precheck_then_block, args=(weak_cancel, weak_handoff, 21), daemon=True)
        weak.start()
        self.assertTrue(weak_handoff.waiting.wait(1))
        weak_cancel.cancel()
        self.assertTrue(weak.is_alive(), "weak send should remain blocked after cancel")
        weak_handoff.make_receiver_ready()  # Release the intentional weak block.
        weak.join(1)
        self.assertFalse(weak.is_alive())
        self.assertEqual(weak_handoff.accepted, 42)

        strong_cancel, strong_handoff = CancelSignal(), ResultHandoff()
        result = []
        strong = threading.Thread(target=lambda: result.append(blocking_send_with_cancel(strong_cancel, strong_handoff, 42)), daemon=True)
        strong.start()
        self.assertTrue(strong_handoff.waiting.wait(1))
        strong_cancel.cancel()
        strong.join(1)
        self.assertFalse(strong.is_alive())
        self.assertEqual(result, [False])

        ready_cancel, ready_handoff = CancelSignal(), ResultHandoff()
        ready_handoff.make_receiver_ready()
        self.assertTrue(blocking_send_with_cancel(ready_cancel, ready_handoff, 42))
        ready_cancel.cancel()  # Late cancel does not retract a delivered result.
        self.assertEqual(ready_handoff.accepted, 42)
        self.assertFalse(long_work_check_cancel(strong_cancel, 100000))

        release, entered = threading.Event(), threading.Event()
        downstream = threading.Thread(target=downstream_without_cancellation, args=(release, entered), daemon=True)
        downstream.start()
        self.assertTrue(entered.wait(1))
        self.assertTrue(downstream.is_alive())
        release.set()
        downstream.join(1)
        self.assertFalse(downstream.is_alive())


if __name__ == "__main__":
    unittest.main()

File: examples/java/Case07Test.java
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

final class Case07Test {
    public static void main(String[] args) throws Exception {
        Case07.CancelSignal weakCancel = new Case07.CancelSignal();
        SynchronousQueue<Integer> weakOut = new SynchronousQueue<>();
        CountDownLatch weakEntered = new CountDownLatch(1);
        Thread weak = new Thread(() -> {
            try { Case07.precheckThenBlock(weakCancel, weakOut, 21, weakEntered); }
            catch (InterruptedException e) { throw new AssertionError(e); }
        });
        weak.start();
        if (!weakEntered.await(1, TimeUnit.SECONDS)) throw new AssertionError("weak send not entered");
        weakCancel.cancel();
        if (!weak.isAlive()) throw new AssertionError("weak send should remain blocked");
        if (weakOut.take() != 42) throw new AssertionError("cleanup receiver got wrong result");
        weak.join(1000);
        if (weak.isAlive()) throw new AssertionError("weak cleanup failed");

        Case07.CancelSignal strongCancel = new Case07.CancelSignal();
        CountDownLatch strongEntered = new CountDownLatch(1);
        AtomicBoolean strongResult = new AtomicBoolean(true);
        Thread strong = new Thread(() -> {
            try { strongResult.set(Case07.blockingSendWithCancel(strongCancel, new SynchronousQueue<>(), 42, strongEntered)); }
            catch (InterruptedException e) { throw new AssertionError(e); }
        });
        strong.start();
        if (!strongEntered.await(1, TimeUnit.SECONDS)) throw new AssertionError("strong send not entered");
        strongCancel.cancel();
        strong.join(1000);
        if (strong.isAlive() || strongResult.get()) throw new AssertionError("strong send did not cancel");

        Case07.CancelSignal readyCancel = new Case07.CancelSignal();
        SynchronousQueue<Integer> readyOut = new SynchronousQueue<>();
        CountDownLatch readyEntered = new CountDownLatch(1);
        AtomicInteger delivered = new AtomicInteger();
        Thread receiver = new Thread(() -> {
            try { delivered.set(readyOut.take()); }
            catch (InterruptedException e) { throw new AssertionError(e); }
        });
        receiver.start();
        if (!Case07.blockingSendWithCancel(readyCancel, readyOut, 42, readyEntered))
            throw new AssertionError("ready receiver should complete handoff");
        receiver.join(1000);
        readyCancel.cancel(); // Late cancel does not undo delivery.
        if (receiver.isAlive() || delivered.get() != 42) throw new AssertionError("late cancel changed result");
        if (Case07.longWorkCheckCancel(strongCancel, 100000)) throw new AssertionError("long work ignored cancel");

        CountDownLatch release = new CountDownLatch(1), entered = new CountDownLatch(1);
        Thread downstream = new Thread(() -> {
            try { Case07.downstreamWithoutCancellation(release, entered); }
            catch (InterruptedException e) { throw new AssertionError(e); }
        });
        downstream.start();
        if (!entered.await(1, TimeUnit.SECONDS) || !downstream.isAlive())
            throw new AssertionError("downstream did not block");
        release.countDown();
        downstream.join(1000);
        if (downstream.isAlive()) throw new AssertionError("downstream cleanup failed");
    }
}
